### PR TITLE
Added Catboost with hyper parameter tuning

### LINKED ISSUE #95: Add Catboost with hyper parameter tuning
### Prerequisites

- [X] I am an Open Source Contributor
- [X] I checked to make sure that this issue has not already been filed

### Description

Implementing the CatBoost model and compare the results with other pre-exisiting models in the repo and do the hyper parameter tuning using Optuna library.

I am a GSSOC'23 contributor.

### Code of Conduct

- [X] I follow [Contributing Guidelines](https://github.com/Rakesh9100/ML-Project-Drug-Review-Dataset/blob/main/.github/CONTRIBUTING_GUIDELINES.md) of this project.

### PR DESCRIPTION
**Pull request by Shubhra-31**
# Fixes Issue🛠️

Closes #95 

# Description👨‍💻 

CatBoost implementation using hyper-parameter tuning using Optuna.The optuna took over 10 hours to run and hence the submission was a little late.

# Type of change📄

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# How this has been tested✅

# Checklist✅ 

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I am an Open Source Contributor

# Screenshots/GIF📷

Test and training results
image

Scatter plot of predicted test values 
testing

Scatter plot of predicted training values
training

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -36,15 +38,10 @@
     "date": "string",
     "usefulCount": "int16",
 }
-
-train_df = pd.read_csv(
-    r"datasets\drugsComTrain_raw.tsv", sep="\t", quoting=2, dtype=dtypes
-)
-
+train_df = pd.read_csv(r"ML-Project-Drug-Review-Dataset\datasets\drugsComTrain_raw.tsv", sep="\t", quoting=2, dtype=dtypes)
```
Comment (Rakesh9100) [inline main.py]: Please correct the file path as of before @Shubhra-31

File: main.py [hunk 2]
```diff
@@ -36,15 +38,10 @@
...
+print(train_df.head())
```
Comment (Rakesh9100) [inline main.py]: Delete this line @Shubhra-31

File: main.py [hunk 3]
```diff
@@ -36,15 +38,10 @@
...
 train_df = train_df.sample(frac=0.8, random_state=42)
-test_df = pd.read_csv(
-    r"datasets\drugsComTest_raw.tsv", sep="\t", quoting=2, dtype=dtypes
-)
+test_df = pd.read_csv(r"ML-Project-Drug-Review-Dataset\datasets\drugsComTest_raw.tsv", sep="\t", quoting=2, dtype=dtypes)
```
Comment (Rakesh9100) [inline main.py]: Correct the file path @Shubhra-31

File: main.py [hunk 4]
```diff
@@ -246,17 +241,6 @@ def decode_html(text):
...
-feature = ["drugName", "condition", "rating", "usefulCount"]
-pd.plotting.scatter_matrix(train_imp[feature])
-plt.suptitle("Scatter Matrix For Training DataSet")
-plt.show()
-plt.title("Drug Name Histogram (Training Dataset)")
-plt.hist(train_imp["drugName"], bins=50)
-plt.xlabel("Bins")
-plt.ylabel("Drug Name")
-plt.show()
-
```
Comment (Rakesh9100) [inline main.py]: Dont delete these plots code @Shubhra-31

**Comment by Rakesh9100** [comment]
Pls provide model name as well in the title of both the graphs @Shubhra-31

**Comment by Rakesh9100** [comment]
Pls provide model name as well in the title of both the graphs for clear understanding @Shubhra-31

**Comment by Shubhra-31** [comment]
Done the changes, kindy check @Rakesh9100

**Comment by Rakesh9100** [comment]
Pls run `black main.py` on your editor and then commit to pass the linting check workflow @Shubhra-31

**Comment by Shubhra-31** [comment]
please check @Rakesh9100

**Comment by Rakesh9100** [comment]
> Pls run `black main.py` on your editor and then commit to pass the linting check workflow @Shubhra-31

Already told you to do this pls @Shubhra-31

**Comment by Shubhra-31** [comment]
Could you explain what exactly I have to do ? I am not sure what to do.

**Comment by Rakesh9100** [comment]
Open the project folder in your local editor i.e. Visual Studio and then open the terminal and run `black main.py` and file will be updated with indentation changes and then do the commit @Shubhra-31

**Comment by Shubhra-31** [comment]
Please check @Rakesh9100

**Comment by Rakesh9100** [comment]
You can see the lint check has been failed. Pls follow <https://github.com/Rakesh9100/ML-Project-Drug-Review-Dataset/pull/104#issuecomment-1595844316> properly or connect me on discord. @Shubhra-31

**Pull request merged**